distributed/batched.py

```diff
@@ -38,87 +39,76 @@ class BatchedSend(object):
+            with ignoring(gen.TimeoutError):
+                yield self.waker.wait(self.next_deadline)
+                self.waker.clear()
+            if not self.buffer:
+                # Nothing to send
+                self.next_deadline = None
+                continue
+            if (self.next_deadline is not None and
+                self.loop.time() < self.next_deadline):
+                # Send interval not expired yet
+                continue
+            payload, self.buffer = self.buffer, []
             self.batch_count += 1
-            self.last_send = write(self.stream, payload)
-        except Exception as e:
-            logger.exception(e)
-            raise
+            try:
+                yield write(self.stream, payload)
+            except Exception:
+                logger.exception("Error in batched write")
+                break
+            self.next_deadline = self.loop.time() + self.interval
 
-    @gen.coroutine
-    def _write(self, payload):
-        yield gen.sleep(0)
-        yield write(self.stream, payload)
+        self.stopped.set()
 
     def send(self, msg):
-        """ Send a message to the other side
+        """ Schedule a message for sending to the other side
 
         This completes quickly and synchronously
         """
-        try:
-            self.message_count += 1
-            if self.stream is None:  # not yet started
-                self.buffer.append(msg)
-                return
-
-            if self.stream._closed:
-                raise StreamClosedError()
+        if self.stream is not None and self.stream._closed:
+            raise StreamClosedError()
 
-            if self.buffer:
-                self.buffer.append(msg)
-                return
-
-            # If we're new and early,
-            now = default_timer()
-            if (now < self.last_transmission + self.interval
-                or not self.last_send._done):
-                self.buffer.append(msg)
-                self.loop.add_callback(self.send_next)
-                return
-
-            self.buffer.append(msg)
-            self.loop.add_callback(self.send_next, wait=False)
-        except StreamClosedError:
-            raise
-        except Exception as e:
-            logger.exception(e)
+        self.message_count += 1
+        self.buffer.append(msg)
+        self.waker.set()
 
     @gen.coroutine
     def close(self, ignore_closed=False):
         """ Flush existing messages and then close stream """
+        if self.stream is None:
+            return
+        self.please_stop = True
+        self.waker.set()
+        yield self.stopped.wait()
         try:
-            if self.stream._write_buffer:
-                yield self.last_send
             if self.buffer:
+                if self.next_deadline is not None:
+                    delay = self.next_deadline - self.loop.time()
+                    if delay > 0:
+                        yield gen.sleep(delay)
```

Comment:
It mirrors the `yield self.last_send` that was here previously. Perhaps I'm misunderstanding the intent :-)